Regenerate third-party licenses on trunk pushes

andyfeller opened the pull request:
Fixes #11270

This commit refactors the work done in #11047 of blocking pull requests for manual `third-party` license updates to having GitHub Actions automatically update it on pushes to `trunk`.

This will allow maintainers to streamline Dependabot PR reviews while reducing contributor friction when changing dependencies.

## Demo

1. Demonstrating running GitHub Actions workflow that is canceled by newer workflow run

   <img width="5344" height="3054" alt="Screenshot 2025-08-01 at 3 33 10 PM" src="https://github.com/user-attachments/assets/4fbf5d2b-9d08-4d85-8fb8-924f05d08332" />

1. Demonstrating multiple GitHub Actions workflows being canceled by newer workflow run

   <img width="5344" height="3054" alt="Screenshot 2025-08-01 at 3 32 58 PM" src="https://github.com/user-attachments/assets/432dc34c-fd89-4d47-9315-8f62a01927e5" />

1. Demonstrating updates to third-party license documentation

   <img width="5344" height="3054" alt="Screenshot 2025-08-01 at 3 32 49 PM" src="https://github.com/user-attachments/assets/eaac886a-f973-43b1-a316-ca03cc36824c" />

andyfeller committed:
Regenerate third-party licenses on trunk pushes

Fixes #11270

This commit refactors the work done in #11047 of blocking pull requests for manual `third-party` license updates to having GitHub Actions automatically update it on pushes to `trunk`.

This will allow maintainers to streamline Dependabot PR reviews while reducing contributor friction when changing dependencies.

BagToad commented on .github/workflows/third-party-licenses.yml:
```diff
@@ -0,0 +1,49 @@
+            git add third-party third-party-licenses.*.md
+            git commit -m "Generate licenses - $GITHUB_SHA"
+            git pull
+            git push origin
```
This isn't an issue because the workflow uses concurrency groups

BagToad commented on .github/workflows/third-party-licenses.yml:
```diff
@@ -0,0 +1,49 @@
+  # This job is responsible for updating the third-party license reports and source code.
+  # It should be safe to cancel as the latest version of `go.mod` should be checked in.
+  regenerate-licenses:
+    runs-on: ubuntu-latest
+    concurrency:
+      group: ${{ github.workflow }}
+      cancel-in-progress: true
+    steps:
+      - name: Check out code
+        uses: actions/checkout@v4
+        with:
+          token: ${{ secrets.AUTOMATION_TOKEN }}
+
+      - name: Set up Go
+        uses: actions/setup-go@v5
+        with:
+          go-version-file: 'go.mod'
+
+      - name: Regenerate licenses
+        run: |
+          export GOROOT=$(go env GOROOT)
+          export PATH=${GOROOT}/bin:$PATH
+          go install github.com/google/go-licenses@5348b744d0983d85713295ea08a20cca1654a45e
+          make licenses
+          git diff
+
+      - name: Commit and push changes
+        run: |
+          if git diff --exit-code; then
+            echo "No third-party license changes to commit"
+          else
+            git config --local user.name "github-actions[bot]"
+            git config --local user.email "41898282+github-actions[bot]@users.noreply.github.com"
+            git add third-party third-party-licenses.*.md
+            git commit -m "Generate licenses - $GITHUB_SHA"
+            git pull
+            git push origin
+          fi
```
Yep, this is a good call 👍

andyfeller replied:
@BagToad : since we're leveraging the automation token here, I'd assume we'd have the `permissions` block empty since everything is tied to the token, yeah?

BagToad replied:
> @BagToad : since we're leveraging the automation token here, I'd assume we'd have the `permissions` block empty since everything is tied to the token, yeah?

That works 👍 

Although, we might want to consider using the GITHUB_TOKEN instead 🤔 I think we could very well do so since we don't actually care for our commit to kickoff other workflows (correct me if I'm wrong about that please 🙏 ).

I don't have a strong opinion since this is running in the isolation of `trunk` only. It can use whatever token we feel is most appropriate 👍 Definitely agree that seeing `permissions` to `{}` is the way to go if you want to keep using the automation token.

BagToad commented on .github/workflows/third-party-licenses.yml:
```diff
@@ -0,0 +1,49 @@
+      - name: Regenerate licenses
+        run: |
+          export GOROOT=$(go env GOROOT)
+          export PATH=${GOROOT}/bin:$PATH
```
Fair point I guess, but I don't think that explanation is correct. This is just the latest version, and pinning the commit sha is good practice.

BagToad commented on .github/workflows/third-party-licenses.yml:
```diff
@@ -0,0 +1,49 @@
+      - ".github/licenses.tmpl"
+      - "script/licenses*"
+jobs:
+  # This job is responsible for updating the third-party license reports and source code.
+  # It should be safe to cancel as the latest version of `go.mod` should be checked in.
+  regenerate-licenses:
+    runs-on: ubuntu-latest
+    concurrency:
+      group: ${{ github.workflow }}
+      cancel-in-progress: true
```
Note: There's some nuance here that that might be good to understand.

What I'm referring to is that concurrency groups **[do not have a guaranteed order](https://docs.github.com/en/actions/how-tos/writing-workflows/choosing-when-your-workflow-runs/control-the-concurrency-of-workflows-and-jobs)**:

> Ordering is not guaranteed for jobs or workflow runs using concurrency groups. Jobs or workflow runs in the same concurrency group are handled in an arbitrary order.

This means that if more than one workflow is running close together, you may get the **older** workflow pre-empting the **newer** workflow.

Thinking about what the impact this would have on this particular workflow 🤔 💭 

`actions/checkout` likely checks out the commit sha that triggered the workflow. This will be the commit we run the license generation against. If we run the license generation against an older commit **after** running it on a new commit, I suspect that this workflow may end up **reversing the license updates made by the previous workflow that ran on a newer commit**.

Big point to know here is that this only matters if pushes to trunk happen in short succession. I've been told this can be up to 15 minutes between workflows, but really it depends on many factors like workflow orchestration delays and how long the workflow itself takes to run. In practice I suspect we're looking at problems if multiple PRs are merged within a few minutes of each other.

andyfeller replied:
Very astute call out, @BagToad. 🤔  I'm absolutely worried about concurrency and this adds more concern to that.

Taking all of that in, I can only see a few options for moving forward:

1. **No automated fixes, no changes to Dependabot:** maintainers and contributors manually regenerate `third-party` license information in pull requests, maintainers continue to fix Dependabot PRs

1. **No automated fixes, Dependabot Go updates are grouped up:** maintainers and contributors manually regenerate `third-party` license information in pull requests, maintainers continue to fix Dependabot PRs but have fewer PRs to fix

   ```yml
     - package-ecosystem: "gomod"
       directory: "/"
       schedule:
         interval: "weekly"
       groups:
         go-packages:
           applies-to: version-updates
           patterns: ["*"]
           update-types:
             - "minor"
             - "patch"
   ```

1. **Automate fixing Dependabot PRs, no changes to Dependabot:** maintainers and contributors still manually regenerate `third-party` license information in pull requests, but we avoid concurrent automation impacting `trunk` and limit the security surface of automation doing something unintended

1. **Automate fixing Dependabot PRs, Dependabot Go updates are grouped up:** hybrid of previous 2 options, maintainers and contributor PRs are manually fixed, but fewer Dependabot PRs and they are automatically fixed

BagToad replied:
I've thought more deeply about this and the options, and I'd like to propose a 5th option:

Keep the `on: push` workflow here, but instead of checking out the ref that triggered the PR (which is the default `actions/checkout` behavior), we explicitly always checkout the HEAD of `trunk` by passing `ref` into `actions/checkout`. We also keep the `cancel-in-progress: true` since then it doesn't really matter what order they run in, they'll both perform the same operation of cloning `trunk` and running the license check against it.

andyfeller committed:
Update permissions and events for workflow

This commit makes a few notable changes:

1. Use the GitHub Actions automatic token for committing changes
2. Include workflow file in paths to trigger workflow
3. Checkout the default branch explicitly

BagToad reviewed (approved):
Looks great to me ✨ Thanks for working through the concurrency concern with me and considering my proposal 🙇 I'm glad we were able to push through the blockers with this one and figure it out. Hopefully this does the trick🤞

ftgg0154-ui commented:
> gggjk
